File: benchmarks/server/src/sirannon_bench/core/workloads.py
"""The workload catalogue both engines run identically.

Each workload names its schema for each dialect, a seed that fills the table, and a weighted set
of operations. Operations carry SQL for both dialects written with ``?`` placeholders; each
driver renders the placeholder in its own style. Reads, writes, and read-modify-writes are drawn
by weight, so a run reproduces the same operation stream given the same seed.

The set is the standard OLTP core: point-select, bulk-insert, and batch-update as the atomic
operations, YCSB A/B/C/F as the key-value mix, and a TPC-C-shaped transaction mix. YCSB D
(read-latest) and E (short-range-scan) are left out because D needs insert-driven key growth
during the run and E needs ordered-key range scans on a separately loaded table; running a
stated subset is documented, accepted practice.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

from .rng import SeededRng, ZipfianGenerator

YCSB_SCHEMA = (
    "CREATE TABLE IF NOT EXISTS usertable ("
    "ycsb_key TEXT PRIMARY KEY, "
    "field0 TEXT, field1 TEXT, field2 TEXT, field3 TEXT, field4 TEXT, "
    "field5 TEXT, field6 TEXT, field7 TEXT, field8 TEXT, field9 TEXT)"
)

MICRO_SCHEMA_SQLITE = (
    "CREATE TABLE IF NOT EXISTS users ("
    "id INTEGER PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL, age INTEGER NOT NULL, bio TEXT)"
)

MICRO_SCHEMA_POSTGRES = (
    "CREATE TABLE IF NOT EXISTS users ("
    "id INTEGER PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL, age INTEGER NOT NULL, bio TEXT)"
)

TPCC_SCHEMA_SQLITE = (
    "CREATE TABLE IF NOT EXISTS customers ("
    "id INTEGER PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL, balance REAL NOT NULL);"
    "CREATE TABLE IF NOT EXISTS products ("
    "id INTEGER PRIMARY KEY, name TEXT NOT NULL, price REAL NOT NULL, stock INTEGER NOT NULL);"
    "CREATE TABLE IF NOT EXISTS orders ("
    "id INTEGER PRIMARY KEY, customer_id INTEGER NOT NULL, total REAL NOT NULL, "
    "status TEXT NOT NULL, created_at TEXT NOT NULL)"
)

TPCC_SCHEMA_POSTGRES = (
    "CREATE TABLE IF NOT EXISTS customers ("
    "id INTEGER PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL, balance REAL NOT NULL);"
    "CREATE TABLE IF NOT EXISTS products ("
    "id INTEGER PRIMARY KEY, name TEXT NOT NULL, price REAL NOT NULL, stock INTEGER NOT NULL);"
    "CREATE TABLE IF NOT EXISTS orders ("
    "id SERIAL PRIMARY KEY, customer_id INTEGER NOT NULL, total REAL NOT NULL, "
    "status TEXT NOT NULL, created_at TEXT NOT NULL)"
)

_FIRST_NAMES = ["Alice", "Bob", "Carol", "Dave", "Eve", "Frank", "Grace", "Heidi", "Ivan", "Judy"]
_LAST_NAMES = ["Smith", "Johnson", "Williams", "Brown", "Jones", "Garcia", "Miller", "Davis", "Rodriguez", "Martinez"]
_FIXED_TIMESTAMP = "2026-01-01T00:00:00.000Z"


@dataclass(frozen=True)
class OperationContext:
    rng: SeededRng
    zipf: ZipfianGenerator
    data_size: int


@dataclass(frozen=True)
class Operation:
    name: str
    weight: float
    kind: str  # "read", "write", or "rmw"
    sqlite_sql: str
    postgres_sql: str
    params: Callable[[OperationContext], list]
    write_sqlite_sql: str | None = None
    write_postgres_sql: str | None = None


@dataclass(frozen=True)
class SeedTable:
    table: str
    columns: tuple[str, ...]
    rows: list[list[object]]


@dataclass(frozen=True)
class Workload:
    name: str
    category: str
    tables: tuple[str, ...]
    sqlite_schema: str
    postgres_schema: str
    seed: Callable[[SeededRng, int], list[SeedTable]]
    operations: tuple[Operation, ...]


def _user_row(rng: SeededRng, row_id: int) -> list[object]:
    name = f"{_FIRST_NAMES[row_id % len(_FIRST_NAMES)]} {_LAST_NAMES[row_id % len(_LAST_NAMES)]}"
    email = f"user{row_id}@example.com"
    age = 18 + (row_id % 62)
    bio = f"Bio for user {row_id}: {rng.text(50)}"
    return [row_id, name, email, age, bio]


def _ycsb_row(rng: SeededRng, key: str) -> list[object]:
    return [key, *[rng.text(100) for _ in range(10)]]


def _customer_row(row_id: int) -> list[object]:
    return [row_id, f"Customer {row_id}", f"customer{row_id}@store.com", 10000.0]


def _product_row(rng: SeededRng, row_id: int) -> list[object]:
    price = round((10 + rng.fraction() * 490) * 100) / 100
    return [row_id, f"Product {row_id}", price, 1000]


def _user_seed(rng: SeededRng, data_size: int) -> list[SeedTable]:
    rows = [_user_row(rng, i + 1) for i in range(data_size)]
    return [SeedTable("users", ("id", "name", "email", "age", "bio"), rows)]


def _empty_seed(rng: SeededRng, data_size: int) -> list[SeedTable]:
    return []


def _ycsb_seed(rng: SeededRng, data_size: int) -> list[SeedTable]:
    columns = ("ycsb_key", *[f"field{i}" for i in range(10)])
    rows = [_ycsb_row(rng, f"user{i}") for i in range(data_size)]
    return [SeedTable("usertable", columns, rows)]


def _tpcc_seed(rng: SeededRng, data_size: int) -> list[SeedTable]:
    customers = [_customer_row(i + 1) for i in range(data_size)]
    product_count = min(data_size, 1000)
    products = [_product_row(rng, i + 1) for i in range(product_count)]
    return [
        SeedTable("customers", ("id", "name", "email", "balance"), customers),
        SeedTable("products", ("id", "name", "price", "stock"), products),
    ]


_YCSB_READ = "SELECT * FROM usertable WHERE ycsb_key = ?"
_YCSB_UPDATE = "UPDATE usertable SET field0 = ? WHERE ycsb_key = ?"


def _ycsb_key(ctx: OperationContext) -> str:
    return f"user{ctx.zipf.next(ctx.rng)}"


def _ycsb_value(ctx: OperationContext) -> str:
    return f"value_{ctx.rng.below(1_000_000)}"


def _ycsb_read_op(weight: float) -> Operation:
    return Operation(
        name="read",
        weight=weight,
        kind="read",
        sqlite_sql=_YCSB_READ,
        postgres_sql=_YCSB_READ,
        params=lambda ctx: [_ycsb_key(ctx)],
    )


def _ycsb_update_op(weight: float) -> Operation:
    return Operation(
        name="update",
        weight=weight,
        kind="write",
        sqlite_sql=_YCSB_UPDATE,
        postgres_sql=_YCSB_UPDATE,
        params=lambda ctx: [_ycsb_value(ctx), _ycsb_key(ctx)],
    )


def _ycsb_rmw_op(weight: float) -> Operation:
    return Operation(
        name="read-modify-write",
        weight=weight,
        kind="rmw",
        sqlite_sql=_YCSB_READ,
        postgres_sql=_YCSB_READ,
        write_sqlite_sql=_YCSB_UPDATE,
        write_postgres_sql=_YCSB_UPDATE,
        params=lambda ctx: [_ycsb_key(ctx), _ycsb_value(ctx)],
    )


def _ycsb_workload(name: str, operations: tuple[Operation, ...]) -> Workload:
    return Workload(
        name=name,
        category="ycsb",
        tables=("usertable",),
        sqlite_schema=YCSB_SCHEMA,
        postgres_schema=YCSB_SCHEMA,
        seed=_ycsb_seed,
        operations=operations,
    )


def _point_select_params(ctx: OperationContext) -> list:
    return [ctx.zipf.next(ctx.rng) + 1]


def _bulk_insert_params(ctx: OperationContext) -> list:
    row_id = ctx.rng.below(ctx.data_size) + 1
    return _user_row(ctx.rng, row_id)


def _batch_update_params(ctx: OperationContext) -> list:
    row_id = ctx.zipf.next(ctx.rng) + 1
    return [ctx.rng.below(80) + 18, row_id]


def _new_order_params(ctx: OperationContext) -> list:
    customer_id = ctx.rng.below(ctx.data_size) + 1
    total = round(ctx.rng.fraction() * 500 * 100) / 100
    return [customer_id, total, "pending", _FIXED_TIMESTAMP]


def _payment_params(ctx: OperationContext) -> list:
    amount = round(ctx.rng.fraction() * 100 * 100) / 100
    customer_id = ctx.rng.below(ctx.data_size) + 1
    return [amount, customer_id]


def build_workloads() -> dict[str, Workload]:
    catalogue: list[Workload] = [
        Workload(
            name="point-select",
            category="micro",
            tables=("users",),
            sqlite_schema=MICRO_SCHEMA_SQLITE,
            postgres_schema=MICRO_SCHEMA_POSTGRES,
            seed=_user_seed,
            operations=(
                Operation(
                    name="point-select",
                    weight=1.0,
                    kind="read",
                    sqlite_sql="SELECT * FROM users WHERE id = ?",
                    postgres_sql="SELECT * FROM users WHERE id = ?",
                    params=_point_select_params,
                ),
            ),
        ),
        Workload(
            name="bulk-insert",
            category="micro",
            tables=("users",),
            sqlite_schema=MICRO_SCHEMA_SQLITE,
            postgres_schema=MICRO_SCHEMA_POSTGRES,
            seed=_empty_seed,
            operations=(
                Operation(
                    name="bulk-insert",
                    weight=1.0,
                    kind="write",
                    sqlite_sql="INSERT OR REPLACE INTO users (id, name, email, age, bio) VALUES (?, ?, ?, ?, ?)",
                    postgres_sql=(
                        "INSERT INTO users (id, name, email, age, bio) VALUES (?, ?, ?, ?, ?) "
                        "ON CONFLICT (id) DO UPDATE SET name = EXCLUDED.name"
                    ),
                    params=_bulk_insert_params,
                ),
            ),
        ),
        Workload(
            name="batch-update",
            category="micro",
            tables=("users",),
            sqlite_schema=MICRO_SCHEMA_SQLITE,
            postgres_schema=MICRO_SCHEMA_POSTGRES,
            seed=_user_seed,
            operations=(
                Operation(
                    name="batch-update",
                    weight=1.0,
                    kind="write",
                    sqlite_sql="UPDATE users SET age = ? WHERE id = ?",
                    postgres_sql="UPDATE users SET age = ? WHERE id = ?",
                    params=_batch_update_params,
                ),
            ),
        ),
        _ycsb_workload("ycsb-a", (_ycsb_read_op(0.5), _ycsb_update_op(0.5))),
        _ycsb_workload("ycsb-b", (_ycsb_read_op(0.95), _ycsb_update_op(0.05))),
        _ycsb_workload("ycsb-c", (_ycsb_read_op(1.0),)),
        _ycsb_workload("ycsb-f", (_ycsb_read_op(0.5), _ycsb_rmw_op(0.5))),
        Workload(
            name="tpc-c-derived",
            category="oltp",
            tables=("orders", "products", "customers"),
            sqlite_schema=TPCC_SCHEMA_SQLITE,
            postgres_schema=TPCC_SCHEMA_POSTGRES,
            seed=_tpcc_seed,
            operations=(
                Operation(
                    name="new-order",
                    weight=0.5,
                    kind="write",
                    sqlite_sql="INSERT INTO orders (customer_id, total, status, created_at) VALUES (?, ?, ?, ?)",
                    postgres_sql="INSERT INTO orders (customer_id, total, status, created_at) VALUES (?, ?, ?, ?)",
                    params=_new_order_params,
                ),
                Operation(
                    name="payment",
                    weight=0.5,
                    kind="write",
                    sqlite_sql="UPDATE customers SET balance = balance - ? WHERE id = ?",
                    postgres_sql="UPDATE customers SET balance = balance - ? WHERE id = ?",
                    params=_payment_params,
                ),
            ),
        ),
    ]
    return {workload.name: workload for workload in catalogue}


def pick_operation(rng: SeededRng, operations: tuple[Operation, ...]) -> Operation:
    draw = rng.fraction()
    cumulative = 0.0
    for operation in operations:
        cumulative += operation.weight
        if draw < cumulative:
            return operation
    return operations[-1]

File: benchmarks/server/driver/src/workloads.ts
// The workload catalogue both engines run identically. Each workload names its schema for each
// dialect, a seed that fills the table, and a weighted set of operations. Operations carry SQL for
// both dialects written with `?` placeholders; each driver renders the placeholder in its own
// style. Reads, writes, and read-modify-writes are drawn by weight, so a run reproduces the same
// operation stream given the same seed.
//
// The set is the standard OLTP core: point-select, bulk-insert, and batch-update as the atomic
// operations, YCSB A/B/C/F as the key-value mix, and a TPC-C-shaped transaction mix. YCSB D
// (read-latest) and E (short-range-scan) are left out because D needs insert-driven key growth
// during the run and E needs ordered-key range scans on a separately loaded table; running a
// stated subset is documented, accepted practice.

import type { SeededRng, ZipfianGenerator } from './rng.ts'

export interface OperationContext {
  rng: SeededRng
  zipf: ZipfianGenerator
  dataSize: number
}

export type OperationKind = 'read' | 'write' | 'rmw'

export interface Operation {
  name: string
  weight: number
  kind: OperationKind
  sqliteSql: string
  postgresSql: string
  params: (ctx: OperationContext) => unknown[]
  writeSqliteSql?: string
  writePostgresSql?: string
}

export interface SeedTable {
  table: string
  columns: string[]
  rows: Iterable<unknown[]>
}

export interface Workload {
  name: string
  category: string
  tables: string[]
  sqliteSchema: string
  postgresSchema: string
  seed: (rng: SeededRng, dataSize: number) => SeedTable[]
  operations: Operation[]
}

const YCSB_SCHEMA =
  'CREATE TABLE IF NOT EXISTS usertable (' +
  'ycsb_key TEXT PRIMARY KEY, ' +
  'field0 TEXT, field1 TEXT, field2 TEXT, field3 TEXT, field4 TEXT, ' +
  'field5 TEXT, field6 TEXT, field7 TEXT, field8 TEXT, field9 TEXT)'

const MICRO_SCHEMA =
  'CREATE TABLE IF NOT EXISTS users (' +
  'id INTEGER PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL, age INTEGER NOT NULL, bio TEXT)'

const TPCC_SCHEMA_SQLITE =
  'CREATE TABLE IF NOT EXISTS customers (' +
  'id INTEGER PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL, balance REAL NOT NULL);' +
  'CREATE TABLE IF NOT EXISTS products (' +
  'id INTEGER PRIMARY KEY, name TEXT NOT NULL, price REAL NOT NULL, stock INTEGER NOT NULL);' +
  'CREATE TABLE IF NOT EXISTS orders (' +
  'id INTEGER PRIMARY KEY, customer_id INTEGER NOT NULL, total REAL NOT NULL, ' +
  'status TEXT NOT NULL, created_at TEXT NOT NULL)'

const TPCC_SCHEMA_POSTGRES =
  'CREATE TABLE IF NOT EXISTS customers (' +
  'id INTEGER PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL, balance REAL NOT NULL);' +
  'CREATE TABLE IF NOT EXISTS products (' +
  'id INTEGER PRIMARY KEY, name TEXT NOT NULL, price REAL NOT NULL, stock INTEGER NOT NULL);' +
  'CREATE TABLE IF NOT EXISTS orders (' +
  'id SERIAL PRIMARY KEY, customer_id INTEGER NOT NULL, total REAL NOT NULL, ' +
  'status TEXT NOT NULL, created_at TEXT NOT NULL)'

const FIRST_NAMES = ['Alice', 'Bob', 'Carol', 'Dave', 'Eve', 'Frank', 'Grace', 'Heidi', 'Ivan', 'Judy']
const LAST_NAMES = ['Smith', 'Johnson', 'Williams', 'Brown', 'Jones', 'Garcia', 'Miller', 'Davis', 'Rodriguez', 'Martinez']
const FIXED_TIMESTAMP = '2026-01-01T00:00:00.000Z'

function userRow(rng: SeededRng, rowId: number): unknown[] {
  const name = `${FIRST_NAMES[rowId % FIRST_NAMES.length]} ${LAST_NAMES[rowId % LAST_NAMES.length]}`
  const email = `user${rowId}@example.com`
  const age = 18 + (rowId % 62)
  const bio = `Bio for user ${rowId}: ${rng.text(50)}`
  return [rowId, name, email, age, bio]
}

function ycsbRow(rng: SeededRng, key: string): unknown[] {
  const fields: string[] = []
  for (let i = 0; i < 10; i++) {
    fields.push(rng.text(100))
  }
  return [key, ...fields]
}

function customerRow(rowId: number): unknown[] {
  return [rowId, `Customer ${rowId}`, `customer${rowId}@store.com`, 10000.0]
}

function productRow(rng: SeededRng, rowId: number): unknown[] {
  const price = Math.round((10 + rng.fraction() * 490) * 100) / 100
  return [rowId, `Product ${rowId}`, price, 1000]
}

function* userRows(rng: SeededRng, dataSize: number): Generator<unknown[]> {
  for (let i = 0; i < dataSize; i++) {
    yield userRow(rng, i + 1)
  }
}

function userSeed(rng: SeededRng, dataSize: number): SeedTable[] {
  return [{ table: 'users', columns: ['id', 'name', 'email', 'age', 'bio'], rows: userRows(rng, dataSize) }]
}

function emptySeed(): SeedTable[] {
  return []
}

function* ycsbRows(rng: SeededRng, dataSize: number): Generator<unknown[]> {
  for (let i = 0; i < dataSize; i++) {
    yield ycsbRow(rng, `user${i}`)
  }
}

function ycsbSeed(rng: SeededRng, dataSize: number): SeedTable[] {
  const columns = ['ycsb_key']
  for (let i = 0; i < 10; i++) {
    columns.push(`field${i}`)
  }
  return [{ table: 'usertable', columns, rows: ycsbRows(rng, dataSize) }]
}

function* customerRows(dataSize: number): Generator<unknown[]> {
  for (let i = 0; i < dataSize; i++) {
    yield customerRow(i + 1)
  }
}

function* productRows(rng: SeededRng, count: number): Generator<unknown[]> {
  for (let i = 0; i < count; i++) {
    yield productRow(rng, i + 1)
  }
}

function tpccSeed(rng: SeededRng, dataSize: number): SeedTable[] {
  const productCount = Math.min(dataSize, 1000)
  return [
    { table: 'customers', columns: ['id', 'name', 'email', 'balance'], rows: customerRows(dataSize) },
    { table: 'products', columns: ['id', 'name', 'price', 'stock'], rows: productRows(rng, productCount) },
  ]
}

const YCSB_READ = 'SELECT * FROM usertable WHERE ycsb_key = ?'
const YCSB_UPDATE = 'UPDATE usertable SET field0 = ? WHERE ycsb_key = ?'

function ycsbKey(ctx: OperationContext): string {
  return `user${ctx.zipf.next(ctx.rng)}`
}

function ycsbValue(ctx: OperationContext): string {
  return `value_${ctx.rng.below(1_000_000)}`
}

function ycsbReadOp(weight: number): Operation {
  return {
    name: 'read',
    weight,
    kind: 'read',
    sqliteSql: YCSB_READ,
    postgresSql: YCSB_READ,
    params: ctx => [ycsbKey(ctx)],
  }
}

function ycsbUpdateOp(weight: number): Operation {
  return {
    name: 'update',
    weight,
    kind: 'write',
    sqliteSql: YCSB_UPDATE,
    postgresSql: YCSB_UPDATE,
    params: ctx => [ycsbValue(ctx), ycsbKey(ctx)],
  }
}

function ycsbRmwOp(weight: number): Operation {
  return {
    name: 'read-modify-write',
    weight,
    kind: 'rmw',
    sqliteSql: YCSB_READ,
    postgresSql: YCSB_READ,
    writeSqliteSql: YCSB_UPDATE,
    writePostgresSql: YCSB_UPDATE,
    params: ctx => [ycsbKey(ctx), ycsbValue(ctx)],
  }
}

function ycsbWorkload(name: string, operations: Operation[]): Workload {
  return {
    name,
    category: 'ycsb',
    tables: ['usertable'],
    sqliteSchema: YCSB_SCHEMA,
    postgresSchema: YCSB_SCHEMA,
    seed: ycsbSeed,
    operations,
  }
}

function pointSelectParams(ctx: OperationContext): unknown[] {
  return [ctx.zipf.next(ctx.rng) + 1]
}

function bulkInsertParams(ctx: OperationContext): unknown[] {
  const rowId = ctx.rng.below(ctx.dataSize) + 1
  return userRow(ctx.rng, rowId)
}

function batchUpdateParams(ctx: OperationContext): unknown[] {
  const rowId = ctx.zipf.next(ctx.rng) + 1
  return [ctx.rng.below(80) + 18, rowId]
}

function newOrderParams(ctx: OperationContext): unknown[] {
  const customerId = ctx.rng.below(ctx.dataSize) + 1
  const total = Math.round(ctx.rng.fraction() * 500 * 100) / 100
  return [customerId, total, 'pending', FIXED_TIMESTAMP]
}

function paymentParams(ctx: OperationContext): unknown[] {
  const amount = Math.round(ctx.rng.fraction() * 100 * 100) / 100
  const customerId = ctx.rng.below(ctx.dataSize) + 1
  return [amount, customerId]
}

export function buildWorkloads(): Map<string, Workload> {
  const catalogue: Workload[] = [
    {
      name: 'point-select',
      category: 'micro',
      tables: ['users'],
      sqliteSchema: MICRO_SCHEMA,
      postgresSchema: MICRO_SCHEMA,
      seed: userSeed,
      operations: [
        {
          name: 'point-select',
          weight: 1.0,
          kind: 'read',
          sqliteSql: 'SELECT * FROM users WHERE id = ?',
          postgresSql: 'SELECT * FROM users WHERE id = ?',
          params: pointSelectParams,
        },
      ],
    },
    {
      name: 'bulk-insert',
      category: 'micro',
      tables: ['users'],
      sqliteSchema: MICRO_SCHEMA,
      postgresSchema: MICRO_SCHEMA,
      seed: emptySeed,
      operations: [
        {
          name: 'bulk-insert',
          weight: 1.0,
          kind: 'write',
          sqliteSql: 'INSERT OR REPLACE INTO users (id, name, email, age, bio) VALUES (?, ?, ?, ?, ?)',
          postgresSql:
            'INSERT INTO users (id, name, email, age, bio) VALUES (?, ?, ?, ?, ?) ' +
            'ON CONFLICT (id) DO UPDATE SET name = EXCLUDED.name',
          params: bulkInsertParams,
        },
      ],
    },
    {
      name: 'batch-update',
      category: 'micro',
      tables: ['users'],
      sqliteSchema: MICRO_SCHEMA,
      postgresSchema: MICRO_SCHEMA,
      seed: userSeed,
      operations: [
        {
          name: 'batch-update',
          weight: 1.0,
          kind: 'write',
          sqliteSql: 'UPDATE users SET age = ? WHERE id = ?',
          postgresSql: 'UPDATE users SET age = ? WHERE id = ?',
          params: batchUpdateParams,
        },
      ],
    },
    ycsbWorkload('ycsb-a', [ycsbReadOp(0.5), ycsbUpdateOp(0.5)]),
    ycsbWorkload('ycsb-b', [ycsbReadOp(0.95), ycsbUpdateOp(0.05)]),
    ycsbWorkload('ycsb-c', [ycsbReadOp(1.0)]),
    ycsbWorkload('ycsb-f', [ycsbReadOp(0.5), ycsbRmwOp(0.5)]),
    {
      name: 'tpc-c-derived',
      category: 'oltp',
      tables: ['orders', 'products', 'customers'],
      sqliteSchema: TPCC_SCHEMA_SQLITE,
      postgresSchema: TPCC_SCHEMA_POSTGRES,
      seed: tpccSeed,
      operations: [
        {
          name: 'new-order',
          weight: 0.5,
          kind: 'write',
          sqliteSql: 'INSERT INTO orders (customer_id, total, status, created_at) VALUES (?, ?, ?, ?)',
          postgresSql: 'INSERT INTO orders (customer_id, total, status, created_at) VALUES (?, ?, ?, ?)',
          params: newOrderParams,
        },
        {
          name: 'payment',
          weight: 0.5,
          kind: 'write',
          sqliteSql: 'UPDATE customers SET balance = balance - ? WHERE id = ?',
          postgresSql: 'UPDATE customers SET balance = balance - ? WHERE id = ?',
          params: paymentParams,
        },
      ],
    },
  ]
  return new Map(catalogue.map(workload => [workload.name, workload]))
}

export function pickOperation(rng: SeededRng, operations: Operation[]): Operation {
  const draw = rng.fraction()
  let cumulative = 0.0
  for (const operation of operations) {
    cumulative += operation.weight
    if (draw < cumulative) {
      return operation
    }
  }
  return operations[operations.length - 1] as Operation
}
